### PR TITLE
RFD: Long-lived Machine ID local SSH proxy

### PR DESCRIPTION
**Pull request by strideynet**
RFD for the design of a long-lived Machine ID local SSH proxy.

Related to: https://github.com/gravitational/teleport/issues/41977

### CODE REVIEW

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [inline rfd/0172-tbot-ssh-muxer.md:71]: @espadolini do we need to have some way of indicating failure messages ? Or are we happy to just close the connection and log on the serverside.

File: rfd/0172-tbot-ssh-muxer.md
Comment (espadolini) [inline rfd/0172-tbot-ssh-muxer.md:98]: We could set parameters during some `init()` too, but I'm not sure if they'd cover everything and if it changes our decision, ultimately.

File: rfd/0172-tbot-ssh-muxer.md
Comment (espadolini) [inline rfd/0172-tbot-ssh-muxer.md:96]: The actual code is still like 4 or 5 syscalls including `exit()` in both cases so "speed of execution" is not really different I don't think, so I'd say something like 
```suggestion
- Rust has little to no overhead, whilst Go requires spinning up a runtime with multiple OS threads and a garbage collector before any user code gets to run.
```

File: rfd/0172-tbot-ssh-muxer.md
Comment (espadolini) [inline rfd/0172-tbot-ssh-muxer.md:110]: How is this in any way practical, with `ControlMaster`/`ControlPath`? Connections won't get freed until minutes later, and at that point you've basically just rejected the connection.

File: rfd/0172-tbot-ssh-muxer.md
Comment (espadolini) [inline rfd/0172-tbot-ssh-muxer.md:134]: Huh... why?

File: rfd/0172-tbot-ssh-muxer.md
Comment (espadolini) [inline rfd/0172-tbot-ssh-muxer.md:171]: Permissions on unix sockets are always kind of weird and there's no good way to set the permission of a socket as you're creating it except by playing with the umask (which is process-wide, yuck), so we should take care to use a _subdirectory_ with restrictive permissions that contains the socket - at which point the socket permissions don't matter beyond being at least `u+rw`.

**Review by espadolini — COMMENTED** [review]
I'd mention that right now, stopping a significant amount of `tbot proxy ssh` at once (like what happens when you're done with your connections all at once and a minute later the background ssh muxes all terminate at roughly the same time) will trash the linux scheduler _hard_.

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md:135]: I worry a little bit about performance issues introduced by head of line blocking etc or overuse of the shared connection. If we want to do this, we probably need to carve out some time to determine a safe number of downstream connections to upstream connections. If you're confident we can determine a safe value for this immediately then I'm happy to include this in the work.

Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md:135]: > Without connection pooling are we really going to reduce resource consumption?

Even without connection pooling we:

- Significantly reduce the work for the Linux scheduler by reducing the number of tasks/processes that are active
- Reducing the amount of memory consumed by the overhead of each of those tbots. Even with the improvements we've made to the heap size, we're probably still using a significant amount more memory than actually necessary to run run a single connection.

@espadolini  is going to run a test with one upstream connection per downstream connection - that'll give us a comparison point and we can make a firmer decision on whether or not the initial iteration needs to include connection sharing.

By "future improvement", I'm not really envisioning this in the far future - more just trying to tighten the scope on the initial version so we don't bite off more than we can chew. Effective connection pooling/sharing seems non-trivial to me - and if there's a significant improvement without adding it, then it seems to me to make more sense for us to work iteratively to keep complexity low. If we can deliver a significant improvement sooner and without complexity, it reduces the pressure on us and means we can spend the time we need ensure that the connection pooling/sharing is stable.

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md:99]: Yeah - I'm not sure it makes a major difference either way but I think the very fact that we want to tune the GC is not a good sign for Go.

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md:111]: Do you feel that we would be better off removing any form of limitation on the client side ? Or should we be more aggressive and reject the connections rather than holding off ?

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md:172]: At least for `tbot`, we typically operate on the assumption that the user the bot runs as will often not be the same user that is consuming the credentials output by `tbot`. We handled this for the Workload Identity API by chmodding the newly created socket ( which obviously is a little racy) to 777 and then encouraging control by placing more restrictive permissions on the directory which is not recreated on each start.

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md:143]: 👍  I'll add this as one consideration.

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md:135]: Having gone through the figures with Edoardo - I think this is a big enough improvement that we can proceed straight to including a connection cycler.

Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md]: Addressed in [e141c4b](https://github.com/gravitational/teleport/pull/42546/commits/e141c4b06d710f67402f3d49078b7c069643599f)

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md:70]: Yah - I mean really this is just going to be whatever %h and %p is - so proxy templating will occur within the daemon to resolve that. I'll add anote.

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md:86]: I discussed this with Edoardo - since there's no reason we could have a subcommand by default and then users can switch to the more performant standalone if they want. It'll definitely put a dink in the performance but I'm unsure if @espadolini has done a measurement on that.

File: rfd/0172-tbot-ssh-muxer.md
Comment (espadolini) [inline rfd/0172-tbot-ssh-muxer.md:69]:
```suggestion
  a. The request will be encoded in JSON and be followed by a NUL (ASCII/Unicode 0) character.
```

File: rfd/0172-tbot-ssh-muxer.md
Comment (espadolini) [inline rfd/0172-tbot-ssh-muxer.md:79]:
```suggestion
UNIX domain socket, so any advantage gained by the use of a standard protocol
```

File: rfd/0172-tbot-ssh-muxer.md
Comment (espadolini) [reply rfd/0172-tbot-ssh-muxer.md:130]: Hard to say. Rejecting connections would be better, but blocking will let a bunch of short-lived connection through while respecting the concurrency limit, and both options can be good depending on the workload.

Do we need a connection limit?

File: rfd/0172-tbot-ssh-muxer.md
Comment (espadolini) [inline rfd/0172-tbot-ssh-muxer.md:176]: Is this the limit per grpc connection, or global for the whole service?

File: rfd/0172-tbot-ssh-muxer.md
Comment (espadolini) [reply rfd/0172-tbot-ssh-muxer.md:190]: The listening socket should still be in a dedicated subdirectory that we chmod 700 before doing anything else.

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md:130]: Ended up dropping the limit from implementation - I'll remove from RFD.

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md]: [9866425](https://github.com/gravitational/teleport/pull/42546/commits/98664250527be2ab02dd8b914e25d4ab9486b195)

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md]: [9866425](https://github.com/gravitational/teleport/pull/42546/commits/98664250527be2ab02dd8b914e25d4ab9486b195)

File: rfd/0172-tbot-ssh-muxer.md
Comment (espadolini) [inline rfd/0172-tbot-ssh-muxer.md:67]: Characters are sort of at a premium when dealing with unix domain sockets; considering that we're in a tbot output directory for the ssh-multiplexer service already, why not `mux/v1.sock`?

File: rfd/0172-tbot-ssh-muxer.md
Comment (espadolini) [inline rfd/0172-tbot-ssh-muxer.md:72]: What are we gaining with the JSON?

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md:67]: Sounds good to me

Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md]: a082be328b

File: rfd/0172-tbot-ssh-muxer.md
Comment (strideynet) [reply rfd/0172-tbot-ssh-muxer.md:72]: I'm not extremely married to this decision - I went with JSON as it would make it trivial to add additional optional fields in the future as necessary without the need for breaking the API version. If there's some perceived risk with JSON, I'm more than happy to switch it back to just `host:port` terminated with NULL.

**Review by rosstimothy — APPROVED** [review]
LGTM once the outstanding comments from @espadolini are addressed.

File: rfd/0172-tbot-ssh-muxer.md
Comment (espadolini) [reply rfd/0172-tbot-ssh-muxer.md:183]: I'm fine with using `tbot`'s directory ACL management

**Review by timothyb89 — APPROVED** [review]
Looks good to me as well!

**Pull request merged**
